{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing Labels for ML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Required Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "BITOME_KB_PATH = Path('..', 'bitome-kb')\n",
    "sys.path.append(BITOME_KB_PATH.absolute().as_posix())\n",
    "\n",
    "from bitome.core import Bitome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define Global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LOCAL_CACHE_PATH = Path('..', 'local_cache')\n",
    "DATA_PATH = Path('..', 'data')\n",
    "CLASS_DATA_PATH = Path('..', 'data', 'classification')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Pre-Prepared Bitome Knowledgebase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bitome = Bitome.init_from_file(Path(LOCAL_CACHE_PATH, 'bitome.pkl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load ALE SNP Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Details</th>\n",
       "      <th>Type</th>\n",
       "      <th>From</th>\n",
       "      <th>To</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1196220</td>\n",
       "      <td>H366H (CAC→CAT)</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>H</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1196232</td>\n",
       "      <td>T370T (ACC→ACT)</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1196245</td>\n",
       "      <td>L375M (TTA→CTG)</td>\n",
       "      <td>missense</td>\n",
       "      <td>L</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1196247</td>\n",
       "      <td>L375M (TTA→CTG)</td>\n",
       "      <td>missense</td>\n",
       "      <td>L</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1677707</td>\n",
       "      <td>S66* (TCA→TAA)</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>S</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3060146</td>\n",
       "      <td>Q132K (CAG→AAG)</td>\n",
       "      <td>missense</td>\n",
       "      <td>Q</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3471274</td>\n",
       "      <td>G19V (GGC→GTC)</td>\n",
       "      <td>missense</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>4430521</td>\n",
       "      <td>I220V (ATA→GTA)</td>\n",
       "      <td>missense</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1405</td>\n",
       "      <td>S357R (AGC→CGC)</td>\n",
       "      <td>missense</td>\n",
       "      <td>S</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1756445</td>\n",
       "      <td>E250* (GAA→TAA)</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>E</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8297 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Position          Details        Type From    To\n",
       "2    1196220  H366H (CAC→CAT)  synonymous    H     H\n",
       "3    1196232  T370T (ACC→ACT)  synonymous    T     T\n",
       "4    1196245  L375M (TTA→CTG)    missense    L     M\n",
       "5    1196247  L375M (TTA→CTG)    missense    L     M\n",
       "6    1677707   S66* (TCA→TAA)    nonsense    S  None\n",
       "..       ...              ...         ...  ...   ...\n",
       "28   3060146  Q132K (CAG→AAG)    missense    Q     K\n",
       "29   3471274   G19V (GGC→GTC)    missense    G     V\n",
       "33   4430521  I220V (ATA→GTA)    missense    I     V\n",
       "34      1405  S357R (AGC→CGC)    missense    S     R\n",
       "36   1756445  E250* (GAA→TAA)    nonsense    E  None\n",
       "\n",
       "[8297 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# --- SNP frequency by nt information frequency ---\n",
    "ale_data_path = Path(DATA_PATH, 'aledb_muts_public')\n",
    "snp_df = pd.DataFrame(columns=['Position', 'Details'])\n",
    "for mutation_data_file in os.listdir(ale_data_path):\n",
    "    mutation_df = pd.read_csv(\n",
    "        Path(ale_data_path, mutation_data_file),\n",
    "        index_col=None,\n",
    "        usecols=[0, 1, 2, 3, 10],\n",
    "        names=['Mut ID', 'Reference Seq', 'Position', 'Mutation Type', 'Details']\n",
    "    )\n",
    "    # verified that there are no variants of the NC_000913 ref seq tag in any of the files\n",
    "    mutation_df_ecoli = mutation_df[mutation_df['Reference Seq'] == 'NC_000913']\n",
    "    mutation_df_snps = mutation_df_ecoli[mutation_df_ecoli['Mutation Type'] == 'SNP'][[\n",
    "        'Position',\n",
    "        'Details'\n",
    "    ]]\n",
    "    snp_df = pd.concat([snp_df, mutation_df_snps])\n",
    "\n",
    "def snp_type_from_details(snp_row) -> str:\n",
    "    \"\"\"\n",
    "    Parses the details string from an ALEdb file to determine if a SNP is intergenic, synonymous or non-synonymous\n",
    "\n",
    "    :param snp_row: the raw details string to parse\n",
    "    :return str type_of_snp: the resulting SNP type\n",
    "    \"\"\"\n",
    "    details = snp_row.Details\n",
    "    if 'intergenic' in details or 'pseudogene' in details or 'noncoding' in details:\n",
    "        type_of_snp = 'intergenic'\n",
    "    else:\n",
    "        # assume just one\n",
    "        snp_aa_code = list(re.findall(r'[A-Z*][0-9]{1,5}[A-Z*]', details))[0]\n",
    "        if snp_aa_code[0] == snp_aa_code[-1]:\n",
    "            type_of_snp = 'synonymous'\n",
    "        elif snp_aa_code[-1] == '*':\n",
    "            type_of_snp = 'nonsense'\n",
    "        else:\n",
    "            type_of_snp = 'missense'\n",
    "    return type_of_snp\n",
    "\n",
    "def snp_aa_from_details(snp_row, mutated=False) -> str:\n",
    "    \"\"\"\n",
    "    Parses the details string from an ALEdb file to determine the amino acid that a SNP targeted\n",
    "\n",
    "    :param snp_row: the raw details string to parse\n",
    "    :param bool mutated: indicates if the amino acid that was mutated TO should be returned\n",
    "    :return str snp_aa: the AA the SNP targeted\n",
    "    \"\"\"\n",
    "    details = snp_row.Details\n",
    "    if 'intergenic' in details or 'pseudogene' in details or 'noncoding' in details:\n",
    "        snp_aa = None\n",
    "    else:\n",
    "        # assume just one\n",
    "        snp_aa_code = list(re.findall(r'[A-Z*][0-9]{1,5}[A-Z*]', details))[0]\n",
    "        if mutated:\n",
    "            snp_aa = snp_aa_code[-1]\n",
    "        else:\n",
    "            snp_aa = snp_aa_code[0]\n",
    "        if snp_aa == '*':\n",
    "            snp_aa = None\n",
    "    return snp_aa\n",
    "\n",
    "\n",
    "# cleaning up the dataframe to make sure we have integer positions, some extra mutation information\n",
    "snp_df['Position'] = snp_df.apply(lambda row: int(row.Position.replace(',', '')), axis=1)\n",
    "snp_df['Type'] = snp_df.apply(snp_type_from_details, axis=1)\n",
    "snp_df['From'] = snp_df.apply(snp_aa_from_details, axis=1)\n",
    "snp_df['To'] = snp_df.apply(lambda row: snp_aa_from_details(row, mutated=True), axis=1)\n",
    "\n",
    "# import this separate file that has all K12 mutants; this one has midpoints and tech replicates removed, though\n",
    "# so we can use it to filter down the public; this is a little hacky instead of using Patrick's pipeline to\n",
    "# actually just get this filtering done on public ALEdb\n",
    "filtered_mut_df_private = pd.read_pickle(Path(DATA_PATH, 'filtered_K12_mut_df.pkl'))\n",
    "snp_df_private = filtered_mut_df_private[filtered_mut_df_private['Mutation Type'] == 'SNP']\n",
    "filtered_coding_snp_df_private = snp_df_private[snp_df_private['coding'] == 1]\n",
    "\n",
    "# only take the positions in our snp_df that also show up in the filtered db\n",
    "snp_df = snp_df[snp_df['Position'].isin(filtered_coding_snp_df_private['Position'].astype(int))]\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine Gene Labels\n",
    "\n",
    "Just need to do this for ALE SNPs; essentiality is an attribute of the gene objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_coding_genes = bitome.coding_genes\n",
    "rna_coding_genes = [gene for gene in bitome.genes if gene.gene_type in ['tRNA', 'tmRNA', 'antisense_RNA', 'rRNA']]\n",
    "genes_to_use = protein_coding_genes + rna_coding_genes\n",
    "\n",
    "essential_labels = [int(g.essential) for g in genes_to_use]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# ensure that the SNP indices are shifted from 1-indexing to 0-indexing\n",
    "unique_snp_locs = snp_df['Position'].unique() - 1\n",
    "\n",
    "# for each gene, determine if it has a SNP or not\n",
    "snp_labels = []\n",
    "\n",
    "for i, gene in enumerate(genes_to_use):\n",
    "    gene_range = gene.location.start.position, gene.location.end.position\n",
    "    overlap = len(set(range(*gene_range)).intersection(set(unique_snp_locs))) > 0\n",
    "    snp_labels.append(int(overlap))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define and Save Locus Tag to Label Lookups\n",
    "\n",
    "We want to try out some different featurization schemes, so we will implement those in the main ML notebook.\n",
    "\n",
    "BUT we must define our test/lockbox set NOW to avoid data leakage. Let's do an 80/20 train/test split."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "locus_tags = [g.locus_tag for g in genes_to_use]\n",
    "\n",
    "genes_snp_train, genes_snp_test, y_train_snp, y_test_snp = train_test_split(\n",
    "    locus_tags,\n",
    "    snp_labels,\n",
    "    test_size=0.2,\n",
    "    random_state=42,\n",
    "    stratify=snp_labels\n",
    ")\n",
    "\n",
    "genes_ess_train, genes_ess_test, y_train_ess, y_test_ess = train_test_split(\n",
    "    locus_tags,\n",
    "    essential_labels,\n",
    "    test_size=0.2,\n",
    "    random_state=42,\n",
    "    stratify=essential_labels\n",
    ")\n",
    "\n",
    "pd.Series(y_train_snp, name='SNP', index=genes_snp_train).to_csv(Path(CLASS_DATA_PATH, 'snp_train.csv'))\n",
    "pd.Series(y_test_snp, name='SNP', index=genes_snp_test).to_csv(Path(CLASS_DATA_PATH, 'snp_test.csv'))\n",
    "\n",
    "pd.Series(y_train_ess, name='SNP', index=genes_ess_train).to_csv(Path(CLASS_DATA_PATH, 'essential_train.csv'))\n",
    "pd.Series(y_test_ess, name='SNP', index=genes_ess_test).to_csv(Path(CLASS_DATA_PATH, 'essential_test.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
